### PR TITLE
Update claimedtasks and pooltasks to support reloading of workflow actions on individual WorkflowItems

### LINKED ISSUE #3043: Update /claimedtasks and /pooltasks to support reloading of workflow actions on individual WorkflowItems.
**Required by:* https://github.com/DSpace/dspace-angular/issues/721

**Describe the solution you'd like**
As described in https://github.com/DSpace/dspace-angular/issues/721#issuecomment-722402119:

> when making a [request to claim a task](https://github.com/4Science/Rest7Contract/blob/main/pooltasks.md#post-method-single-resource-level) the endpoint should return the new `claimedtask` created while now return only a `204`. The same should be when [returning a task to the pool](https://github.com/4Science/Rest7Contract/blob/main/claimedtasks.md#delete-method).

Additionally:

> add new search method by item for [pooltasks endpoint](https://github.com/4Science/Rest7Contract/blob/main/pooltasks.md) and [claimedtasks endpoint](https://github.com/4Science/Rest7Contract/blob/main/claimedtasks.md) to allow to retrieve the new task created after performing a "Accept", "Reject" or "Return to Pool" action

These changes will also require the REST Contracts to be updated for both `claimedtasks` and `pooltasks` endpoints.  This work was estimated at **16 hours**

### PR DESCRIPTION
**Pull request by Micheleboychuk**
## References
Fixes https://github.com/DSpace/DSpace/issues/3043
Rest Contract: https://github.com/DSpace/Rest7Contract/pull/151

## Description
Two endpoints have been modified in this PR.
The first one is ClaimedTask where three methods have been added,
findAllByItem, findByItem and creatAndRetund.
In the second one is PoolTask findAllByItem, findByItem have been added, and the action method has been removed as its functionality has been moved to the new creatAndRetund method of ClaimedTask

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [x] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [x] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new, third-party dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR modifies the REST API, I've linked to the REST Contract page (or open PR) related to this change.

### CODE REVIEW

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ClaimedTaskRestRepository.java
Comment (tdonohue) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ClaimedTaskRestRepository.java:162]: This should have a ` @PreAuthorize("hasAuthority('AUTHENTICATED')")` to verify that only logged in users can access this endpoint.  If you add this annotation, you can remove the check below for a `currentUser`, as it's no longer necessary.

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/PoolTaskRestRepository.java
Comment (tdonohue) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/PoolTaskRestRepository.java:160]: Again, this should have `@PreAuthorize("hasAuthority('AUTHENTICATED')")`, and you should remove the check for a `currentUser` as it's no longer needed

**Review by tdonohue — CHANGES_REQUESTED** [review]
@Micheleboychuk : The code here looks good overall, but I noticed a few minor things needing fixing (see inline comments below).   Could you also please rebase this so that it runs the GitHub CI tests?  I haven't had a chance to test this yet, but will do so once cleaned up. Thanks!

**Review by tdonohue — APPROVED** [review]
👍 Looks good to me @Micheleboychuk .  Thanks!  I've also reviewed the corresponding Rest Contract and approved it: https://github.com/DSpace/Rest7Contract/pull/151

After reviewing this, I've realized this likely could be `1 Approval` (if @benbosman agrees).  It's mostly ITs -- there's only ~200 lines of code changed.   However, if @benbosman has a chance to give this a quick review or test that's great too!

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ClaimedTaskRestRepository.java
Comment (benbosman) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ClaimedTaskRestRepository.java:147]: Both an incorrect item and no claim tasks for the items result in a 404, the incorrect item should most likely be a 429
Same applies for the other searches below

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ClaimedTaskRestRepository.java
Comment (benbosman) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ClaimedTaskRestRepository.java:267]: Small typing error

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/TaskRestRepositoriesIT.java
Comment (benbosman) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/TaskRestRepositoriesIT.java:734]: Shouldn't this be a 429?

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/TaskRestRepositoriesIT.java
Comment (benbosman) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/TaskRestRepositoriesIT.java:3101]: It seems there are no tests for:
* searching with a UUID not in the workflow (an archived item)
* searching with a UUID in the workflow, without pooled tasks

This seems to be missing for the other search methods as well

**Review by benbosman — CHANGES_REQUESTED** [review]
Thanks @Micheleboychuk for this PR

I've included some small comments inline.
We should also wait with merging this until the Angular PR is ready to merge

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ClaimedTaskRestRepository.java
Comment (tdonohue) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ClaimedTaskRestRepository.java:147]: @benbosman : A 429 is "Too Many Requests".  I don't think that's the error code we'd want here.  Is there a reason why 404 isn't sufficient?  I'll look at other search endpoints to see what they return in this scenario.

Comment (tdonohue) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ClaimedTaskRestRepository.java:147]: @benbosman : We are returning 404s on (at least some) other endpoints of this type.  For example

* `/api/config/workflowdefinitions/search/findByCollection?uuid=[invalidCollection]` : https://github.com/DSpace/DSpace/blob/main/dspace-server-webapp/src/test/java/org/dspace/app/rest/WorkflowDefinitionRestRepositoryIT.java#L256
* `/api/eperson/registration/search/findByToken?token=[invalidToken]`: https://github.com/DSpace/DSpace/blob/main/dspace-server-webapp/src/test/java/org/dspace/app/rest/RegistrationRestRepositoryIT.java#L94

In my opinion, a 404 is sufficient here, unless I'm missing your point.

Comment (benbosman) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ClaimedTaskRestRepository.java:147]: Sorry, my mistake.
I meant 422, in line with https://github.com/DSpace/Rest7Contract/blob/main/vocabularies.md#get-controlled-vocabulary-by-metadata-and-collection

Comment (tdonohue) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ClaimedTaskRestRepository.java:147]: Thanks for clarifying.  I'm OK with switching this to 422... I can see how that might give more information to the client. A 404 also is returned if the endpoint is "Not Found", so in this scenario I can understand that a 404 might be misleading -- the request is *valid*, but the UUID param is not.

@abollini : Do you have a strong opinion here?  I can see both sides, and it seems like we've been inconsistent on whether to return 422 or 404 in this scenario.

Comment (abollini) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ClaimedTaskRestRepository.java:147]: thanks @benbosman @tdonohue to note that. `404` is not appropriate because the endpoint exists, it is just a matter of the result that must be returned or the parameters that have been provided.
There are four possible scenario here
- the item UUID is missing, this should lead to a 400 error as it is a mandatory parameter
- the parameter for the item UUID is invalid (it is a string, an integer, etc.) it should be treated as the previous case 400
- the item is not found, this should return 422
- the item exists but there are no claimed tasks for it, this should return an empty page with a 200 (as there is at least the pagination information in the json, if it was a single element search 204 would be used)

we have already had a detailed discussion around both aspects 404 vs 400 vs 422 and about the 200/204 for empty search but I was able to find previous references only for the first see
https://jira.lyrasis.org/browse/DS-4428
https://github.com/DSpace/Rest7Contract/pull/102

This also mean that the two enpoints noted by @tdonohue that currently return 404 are wrong and should be fixed

Comment (abollini) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ClaimedTaskRestRepository.java:147]: @benbosman when a single result is returned by a search method, null is translated to 204 No Content. This is already supported by our RestResourceController, see https://github.com/DSpace/DSpace/blob/main/dspace-server-webapp/src/main/java/org/dspace/app/rest/RestResourceController.java#L1014

Comment (abollini) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ClaimedTaskRestRepository.java:147]: I have also created https://github.com/DSpace/DSpace/issues/3093 to track the discussion about the other two endpoints

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/TaskRestRepositoriesIT.java
Comment (abollini) [reply dspace-server-webapp/src/test/java/org/dspace/app/rest/TaskRestRepositoriesIT.java:3102]: the requested tests have been introduced in https://github.com/DSpace/DSpace/pull/3054/commits/3b5130c5a41d96ccf485b44a6a4d844b124261d3

**Review by tdonohue — APPROVED** [review]
👍 The code here now looks good to me & it seems to align with the contract at https://github.com/DSpace/Rest7Contract/pull/151

**Review by benbosman — CHANGES_REQUESTED** [review]
Thanks @Micheleboychuk for the updates

The code does indeed seem to align with the contract (I haven't tested it yet, but based on the code review)
I noticed a few use cases not covered by the ITs (I don't expect any issues with the current implementation, but without the ITs a future change may break the functionality)

I will test the code as well since I don't expect there's a need for changes which impact the functionality.

I will keep this set to "Request changes" until the Angular PR is ready to ensure it's not accidentally merged (it would break the workflow)

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/PoolTaskRestRepository.java
Comment (abollini) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/PoolTaskRestRepository.java:172]: I'm adding this reference more for me to avoid to look back to that again and again :) https://github.com/DSpace/DSpace/pull/3054/commits/3b5130c5a41d96ccf485b44a6a4d844b124261d3#diff-fc493b8186aa975a2deded8f3c737ceffaa076fad1161fe0c3d7080246237bcdR4000

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/PoolTaskRestRepository.java
Comment (abollini) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/PoolTaskRestRepository.java:150]: I'm adding this reference more for me to avoid to look back to that again and again :) 
https://github.com/DSpace/DSpace/pull/3054/commits/3b5130c5a41d96ccf485b44a6a4d844b124261d3#diff-fc493b8186aa975a2deded8f3c737ceffaa076fad1161fe0c3d7080246237bcdR3924
and next test

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ClaimedTaskRestRepository.java
Comment (abollini) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ClaimedTaskRestRepository.java:151]: I'm adding this reference more for me to avoid to look back to that again and again :)
3b5130c#diff-fc493b8186aa975a2deded8f3c737ceffaa076fad1161fe0c3d7080246237bcdR3924
and next test

Comment (benbosman) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ClaimedTaskRestRepository.java:151]: @abollini this IT verifies /api/workflow/claimedtask/search/findByItem instead of /api/workflow/pooltasks/search/findAllByItem

Specifically for /api/workflow/pooltasks/search/findAllByItem I didn't see a method which tests for an existing item which is not in the workflow

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/PoolTaskRestRepository.java
Comment (benbosman) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/PoolTaskRestRepository.java:150]: @abollini this IT verifies /api/workflow/claimedtask/search/findByItem instead of /api/workflow/pooltasks/search/findAllByItem

Specifically for /api/workflow/pooltasks/search/findAllByItem I didn't see a method which tests for an existing item which is not in the workflow

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/PoolTaskRestRepository.java
Comment (benbosman) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/PoolTaskRestRepository.java:172]: This test is not an archived item

Comment (Micheleboychuk) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/PoolTaskRestRepository.java:172]: @benbosman thanks for the feedback, the test has been added: https://github.com/DSpace/DSpace/pull/3054/commits/cfa8b0348bd8d5d824102b8a7b75c6bc0a86e2a4#diff-fc493b8186aa975a2deded8f3c737ceffaa076fad1161fe0c3d7080246237bcdR3458

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ClaimedTaskRestRepository.java
Comment (Micheleboychuk) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ClaimedTaskRestRepository.java:151]: @benbosman thanks for the feedback, the test has been added: https://github.com/DSpace/DSpace/pull/3054/commits/cfa8b0348bd8d5d824102b8a7b75c6bc0a86e2a4#diff-fc493b8186aa975a2deded8f3c737ceffaa076fad1161fe0c3d7080246237bcdR3763

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/PoolTaskRestRepository.java
Comment (Micheleboychuk) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/PoolTaskRestRepository.java:150]: @benbosman thanks for the feedback, the test has been added: https://github.com/DSpace/DSpace/pull/3054/commits/cfa8b0348bd8d5d824102b8a7b75c6bc0a86e2a4#diff-fc493b8186aa975a2deded8f3c737ceffaa076fad1161fe0c3d7080246237bcdR3489

**Review by benbosman — DISMISSED** [review]
Thanks, this solved everything for me.

It's approved for me, but I won't mark it as such to make sure it's not accidentally merged prior to https://github.com/DSpace/dspace-angular/pull/978

**Comment by tdonohue** [comment]
As the Angular PR (https://github.com/DSpace/dspace-angular/pull/978) is now ready for merger, I've dismissed @benbosman's `requested changes` since he noted he only selected that to block this PR from being merged until the corresponding Angular PR is ready.  Both will be merged shortly.

**Pull request merged**
